### PR TITLE
automation: Add new requestor Job

### PR DESCRIPTION
**Pull request by pranavsaxena17**
Add a new requestor job to the automation framework, allowing users to
make specifically crafted requests to targets.

### CODE REVIEW

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            } else {
+                name =
+                        msg.getRequestHeader().getMethod()
+                                + msg.getRequestHeader().getURI().toString();
+            }
+            if (requestMap.get("data") != null) {
+                msg.getRequestBody().setBody((String) requestMap.get("data"));
+            }
+            try {
+                httpSender.sendAndReceive(msg);
```
Comment (psiinon) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:136]: Need to make sure that the requests are shown in the Sites tree and History.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            } catch (IOException e) {
+                progress.warn(
+                        Constant.messages.getString(
+                                "automation.error.requestor.badnetwork", this.getName(), name, e));
+            }
+            if (requestMap.get("responseCode") != null) {
```
Comment (psiinon) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:142]: Could store it in an int so you dont have to get it again, or just check that the map contains the key

Comment (psiinon) [reply addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:142]: Oh and dont forget that the user can ignore warnings, in which case the responseCode might not be an int ;)

**Review by psiinon — CHANGES_REQUESTED** [review]
A few changes needed :)

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                        new URI(url, true);
+                    } catch (URIException e) {
+                        progress.error(
+                                Constant.messages.getString(
+                                        "automation.error.requestor.badurl",
+                                        this.getName(),
+                                        url,
+                                        request));
+                    }
+                    String method = (String) requestMap.get("method");
```
[the remark this answers is not in the document]
Comment (pranavsaxena17) [reply addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:75]: @thc202 suggested that there are other methods as well, and the user could make use of custom ones as well so we shouldn't restrict the method because of that.  https://wiki.owasp.org/index.php/Testing_for_HTTP_Methods_and_XST_(OWASP-CM-008)#Arbitrary_HTTP_Methods

**Comment by pranavsaxena17** [comment]
Tweaked

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (thc202) [reply addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:79]: It could still default to GET, one less line for the common case.

**Comment by thc202** [comment]
Could you rebase to pick the Error Prone checks?

**Comment by pranavsaxena17** [comment]
@thc202 Updated

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        this.httpSender = httpSender;
+    }
+
+    @Override
+    public void verifyJobSpecificData(LinkedHashMap<?, ?> jobData, AutomationProgress progress) {
+        Object o = jobData.get("requests");
+        if (o == null) {
+            return;
+        }
+        if (o instanceof ArrayList<?>) {
```
Comment (thc202) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:63]: - It would be better to negate the expression, reduces nesting and it's not necessary to scroll all way down to know what the condition is doing.
 - The `runJob` is casting anyway.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            ArrayList<?> requests = (ArrayList<?>) o;
+            for (Object request : requests) {
+                if (request instanceof LinkedHashMap<?, ?>) {
+                    LinkedHashMap<?, ?> requestMap = (LinkedHashMap<?, ?>) request;
+                    String url = (String) requestMap.get("url");
```
Comment (thc202) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:68]: These casts should be checked, if the user added a list or map by mistake it would lead to class cast exception.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+                    } catch (URIException e) {
+                        progress.error(
+                                Constant.messages.getString(
+                                        "automation.error.requestor.badurl",
+                                        this.getName(),
+                                        url,
+                                        request));
+                    }
+                    String method = (String) requestMap.get("method");
+                    if (method == null) {
```
Comment (thc202) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:80]: Should also check if it's empty.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+    public RequestorJob() {}
+
+    public RequestorJob(HttpSender httpSender) {
+        this.httpSender = httpSender;
+    }
+
+    @Override
+    public void verifyJobSpecificData(LinkedHashMap<?, ?> jobData, AutomationProgress progress) {
+        Object o = jobData.get("requests");
```
Comment (thc202) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:59]: The names should be extracted to constants.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+                } catch (IOException e) {
+                    progress.warn(
+                            Constant.messages.getString(
+                                    "automation.error.requestor.badnetwork",
+                                    this.getName(),
+                                    name,
+                                    e));
+                }
+                persistToHistoryAndSitesTree(msg);
+                if (requestMap.containsKey("responseCode")) {
```
Comment (thc202) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:152]: Could just get it.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+                    Object o = requestMap.get("responseCode");
+                    if (o instanceof Integer) {
+                        int expectedCode = (Integer) o;
+                        int receivedCode = msg.getResponseHeader().getStatusCode();
```
Comment (thc202) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:156]: This code will not be correct if the response was not received, should probably skip this check altogether.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    @Override
+    public String getExtraConfigFileData() {
+        return "    requests:\n"
+                + "      - url:                          # URL of the request to be made\n"
+                + "        method:                       # A non-empty request method\n";
+    }
+
+    private void persistToHistoryAndSitesTree(HttpMessage msg) {
+        final HistoryReference historyRef;
+        final ExtensionHistory extHistory =
```
Comment (thc202) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:180]: Should check it's not null.

Comment (thc202) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:180]: The finals can be removed.

File: addOns/automation/src/main/javahelp/org/zaproxy/addon/automation/resources/help/contents/job-requestor.html [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+<H2>YAML</H2>
+
+<pre>
+  - type: requestor                    # Used to send specific requests to targets
+    parameters:
+    requests:                          # A list of requests to make
+      - url:                           # String: URL of the request to be made
+        method:                        # String: A non-empty request method
+        data:                          # String: Data to send in the request body
+        responseCode:                  # Int: The expected response code against which the actual response code will be matched
```
Comment (thc202) [inline addOns/automation/src/main/javahelp/org/zaproxy/addon/automation/resources/help/contents/job-requestor.html:23]: Better to specify what is required.

File: addOns/automation/src/test/resources/org/zaproxy/addon/automation/resources/template-min.yaml [hunk 12]
```diff
@@ -23,6 +23,11 @@ jobs:
       maxAlertsPerRule: 10             # Int: Maximum number of alerts to raise per rule
       scanOnlyInScope: true            # Bool: Only scan URLs in scope (recommended)
       maxBodySizeInBytesToScan:        # Int: Maximum body size to scan, default: 0 - will scan all messages
+  - type: requestor                    # Used to send specific requests to targets
+    parameters:
+    requests:                          # A list of requests to make
+      - url:                           # String: URL of the request to be made
+        method:                        # String: A non-empty request method
```
Comment (thc202) [inline addOns/automation/src/test/resources/org/zaproxy/addon/automation/resources/template-min.yaml:30]: Given the method is optional it could just have URL as the minimum (IMO).

File: addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+                progress.getErrors().get(0), is(equalTo("!automation.error.requestor.badlist!")));
+    }
+
+    @Test
+    void shouldFailIfBadRequestUrl() {
+        // Given
+        AutomationProgress progress = new AutomationProgress();
+        RequestorJob job = new RequestorJob();
+
+        String yamlStr = "requests:\n" + "- url: Not a url\n" + "  method: POST\n";
```
Comment (thc202) [inline addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java:122]: Since the method is optional it could be removed.

File: addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+
+        // Then
+        assertThat(progress.hasErrors(), is(equalTo(true)));
+        assertThat(progress.hasWarnings(), is(equalTo(false)));
+        assertThat(progress.getErrors().size(), is(equalTo(1)));
+        assertThat(progress.getErrors().get(0), is(equalTo("!automation.error.requestor.badurl!")));
+    }
+
+    @Test
+    void shouldFailIfEmptyRequestMethod() {
```
Comment (thc202) [inline addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java:137]: Should also test with empty string.

File: addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+        assertThat(progress.hasErrors(), is(equalTo(true)));
+        assertThat(progress.hasWarnings(), is(equalTo(false)));
+        assertThat(progress.getErrors().size(), is(equalTo(1)));
+        assertThat(
+                progress.getErrors().get(0),
+                is(equalTo("!automation.error.requestor.emptymethod!")));
+    }
+
+    @Test
+    void shouldWarnIfBadResponseCode() {
```
Comment (thc202) [inline addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java:159]: `Bad` → `Invalid`

File: addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+        extHistory = new ExtensionHistory();
+        given(extensionLoader.getExtension(ExtensionHistory.class)).willReturn(extHistory);
+        Control.initSingletonForTesting(Model.getSingleton(), extensionLoader);
+    }
+
+    @Test
+    void shouldReturnDefaultFields() {
+        // Given
+
+        // When
```
Comment (thc202) [inline addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java:73]: `// Given / When`

File: addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+import org.parosproxy.paros.model.Model;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.network.HttpSender;
+import org.yaml.snakeyaml.Yaml;
+import org.zaproxy.addon.automation.AutomationEnvironment;
+import org.zaproxy.addon.automation.AutomationJob;
+import org.zaproxy.addon.automation.AutomationProgress;
+import org.zaproxy.zap.utils.I18N;
+
+public class RequestorJobUnitTest {
```
Comment (thc202) [inline addOns/automation/src/test/java/org/zaproxy/addon/automation/jobs/RequestorJobUnitTest.java:53]: Remove public modifier.

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java (hunk 4 above)
Comment (thc202) [reply addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:63]: Regarding the last comment, does not worth doing anything in this PR.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

File: addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+    }
+
+    @Override
+    public String getType() {
+        return JOB_NAME;
+    }
+
+    @Override
+    public Order getOrder() {
+        return Order.EXPLORE;
```
Comment (psiinon) [inline addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/RequestorJob.java:238]: Sorry, just realised that this should be FIRST_EXPLORE so that it will appear before the spiders, otherwise it could appear after them which is almost certainly not what most people would want.

**Review by psiinon — CHANGES_REQUESTED** [review]
Just one change needed, otherwise looks good :)

**Review by psiinon — APPROVED** [review]
